scripts/bugzilla_attach.py

```diff
@@ -0,0 +1,222 @@
+
+    # Handle the potential errors
+    while "error" in resp_json:
+        # Using an api key will disable retries, so just output the error message
+        if use_api_key:
+            print(resp_json["message"])
+            break
+        # 300: invalid username or password
+        if resp_json["code"] == 300:
+            print("Incorrect username or password.")
+            bugzilla_username = input("Username (leave blank to exit): ")
+            if bugzilla_username == "":
+                print("Username left blank, exiting")
+                exit()
+            bugzilla_password = getpass(prompt="Password: ")
+            resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+            resp_json = resp.json()
+        # 101: Invalid bug id
+        elif resp_json["code"] == 101:
+            print("Invalid bug id")
+            new_bug_id = input("Enter a new bug id (leave blank to exit): ")
+            if new_bug_id == "":
+                print("ID left blank, exiting")
+                exit()
+            bug_id, valid = try_parse_int(new_bug_id)
+            # Try and see if the new supplied ID is a positive integer
+            while not valid or bug_id <= 0:
+                print("Could not parse bug id as valid, try again")
+                new_bug_id = input("Enter a new bug id (leave blank to exit): ")
+                if new_bug_id == "":
+                    print("ID left blank, exiting")
+                    exit()
+                bug_id, valid = try_parse_int(new_bug_id)
+            resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+            resp_json = resp.json()
+
+
+def try_parse_int(value):
+    """Tries to parse the value as an int"""
+    try:
+        return int(value), True
+    except ValueError:
+        return value, False
+
+def send_data(username, password, bug_id, file_data):
+    """Sends the data to the Bugzilla URL as an attachment"""
+    url = BUGZILLA_URL + 'rest/bug/%s/attachment' % bug_id
+    data = {
+        "login": username,
+        "password": password,
+        "ids": [bug_id],
+        "summary": "Result from must-gather command",
+        "content_type": "application/gzip",
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def send_data_with_api_key(api_key, bug_id, file_data):
+    """Sends the data but uses an API key instead of a username and password"""
+    url = BUGZILLA_URL + 'rest/bug/%s/attachment' % bug_id
+    data = {
+        "api_key": api_key,
+        "ids": [bug_id],
+        "summary": "Result from must-gather command",
+        "content_type": "application/gzip",
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def get_lines(file):
+    """Gets the number of lines in the file handle"""
+    return sum(1 for line in file)
+
+def trim_file(file, num_lines):
+    """Trims the file handle to the number of lines"""
```

Comment:
What exactly do you want to do here?

Reply:
Without trimming the files, the compressed archive is easily over 40MB, while the maximum file size for Bugzilla attachments is 19.5MB. So by taking only at most the last 100,000 lines of each file, I'm trying to reduce the size of the attachment while still preserving the most recent log lines and still going back far enough to be useful.

Reply:
I suggest to use [islice](https://docs.python.org/3/library/itertools.html#itertools.islice). It should simplify your code.

In general I would suggest to use file split instead. You do not know in which part of the file we can find valuable information. When you would use more images to collect data it would take more space so 100k lines maybe too much.

Reply:
The issue with splitting the file is that I'd still need some way to figure out which part to upload. I can split the file and not add the first parts to the archive, but that still has the same issue of which is the important part.
As for whether 100k lines is too much, this is something that can be changed, or maybe it can be calculated based on the sizes of the logs generated.
Alternatively, the script can skip trimming altogether and split the archive into multiple attachments that are sent separately. But even on a simple cluster, an untrimmed archive has so far reached 66MB. I don't know if we want to be uploading potentially dozens of archives.

Reply:
My point is that we may trim important data. Usually valuable information is closed to the end of the log file not the beginning. Compressing and splitting seems to be much better than trimming the files.

Reply:
The trimming so that the _last_ 100,000 lines are kept, not the _first_. Sorry if that was confusing. I effectively implemented `tail -n 100000 file > file`

Reply:
I agree with @pkliczewski. We should not be discarding any data without verifying that it is unneeded.

Reply:
What do you think about trimming based on time? In the worst case we would not be able to upload to bugzilla.

Reply:
Not all files are formatted based on a timestamp. For the largest ones (the node logs), I've successfully been able to trim based on a given amount of time, but I'm not sure I can assume that those are going to be the only massive logs.

Reply:
Let's explore it. I see potential to reuse this code in other places.